core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -99,13 +99,16 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
     // their requests and send them back to the pool for rescheduling (this may happen if "docker" operations
     // fail for example, or a container has aged and was destroying itself when a new request was assigned)
     case r: Run =>
-      // Check if the message is resent from the buffer. Only the first message on the buffer can be resent.
-      val isResentFromBuffer = runBuffer.nonEmpty && runBuffer.dequeueOption.exists(_._1.msg == r.msg)
+      // Checks if the current message is the first in the queue
+      val isFirstInQueue = runBuffer.dequeueOption.exists(_._1.msg == r.msg)
+      // A resend is only valid, if the message is the first one in the queue. Otherwise it is a result of a race condition.
+      val isValidResent = r.fromQueue && isFirstInQueue
```

Comment:
Since you already switched to using a `runFirstMessageOfQueue` method, can we simplify our logic towards this:

1. On getting a run message, we always append to the buffer
2. After appending, we'll run the `runFirstMessageOfQueue` method

I imagine that that way, we don't need the `isFirstInQueue` and `isValidResent` checks, as that guarantees that we always send messages off of a queue.

(The `Run` handler could of course optimize, and bypass the queue if it knows it's empty)

WDYT?

Reply:
I also thought about this after I was finished.
The only concern with this that came into my mind was, that one more message is sent for every activation. But I don't think, that this will be a bottleneck.
I think it's worth this change to get the code more clear.

Reply:
I don’t understand the comment about a resend by a race. I may be miopic here as I’ve read only the comments so far. 

A resend (which happens when the proxy can’t schedule the message) will get queued at the back of the line.

Do you mean race in the sense that the message was simply resent to the queue?